File: mokabaselib/src/main/java/io/moka/mokabaselib/views/SpreadTextView.java
package io.moka.mokabaselib.views;


import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

import java.util.ArrayList;


public class SpreadTextView extends View {

    private TextPaint textPaint;
    private Rect textBoundRect;

    private float sunX;
    private float monX;
    private float tueX;
    private float wedX;
    private float thuX;
    private float friX;
    private float satX;

    private float dayY;
    private boolean isBold = false;

    private ArrayList<String> xDatas = new ArrayList<>();

    public SpreadTextView(Context context) {
        this(context, null);
    }

    public SpreadTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView();
    }

    private void initView() {
        textBoundRect = new Rect();

        textPaint = new TextPaint();
        textPaint.setTextAlign(Paint.Align.CENTER);
        textPaint.setAntiAlias(true);
        textPaint.setColor(0xFF616161);
        if (isBold)
            textPaint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));

        xDatas.add("일");
        xDatas.add("월");
        xDatas.add("화");
        xDatas.add("수");
        xDatas.add("목");
        xDatas.add("금");
        xDatas.add("토");
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        float dayWidth = w / 7.0f;
        dayY = h / 2.0f;

        sunX = dayWidth / 2;
        monX = sunX + dayWidth;
        tueX = monX + dayWidth;
        wedX = tueX + dayWidth;
        thuX = wedX + dayWidth;
        friX = thuX + dayWidth;
        satX = friX + dayWidth;

        float textSize = h / 2.5f;
        textPaint.setTextSize(textSize);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        drawText(canvas, xDatas.get(0), textPaint, sunX, dayY);
        drawText(canvas, xDatas.get(1), textPaint, monX, dayY);
        drawText(canvas, xDatas.get(2), textPaint, tueX, dayY);
        drawText(canvas, xDatas.get(3), textPaint, wedX, dayY);
        drawText(canvas, xDatas.get(4), textPaint, thuX, dayY);
        drawText(canvas, xDatas.get(5), textPaint, friX, dayY);
        drawText(canvas, xDatas.get(6), textPaint, satX, dayY);
    }

    private void drawText(Canvas canvas, String text, TextPaint textPaint, float centerX, float centerY) {
        textPaint.getTextBounds(text, 0, text.length(), textBoundRect);
        canvas.drawText(text, centerX, (int) (centerY + textBoundRect.height() / 2), textPaint);
    }

    public void setXDatas(ArrayList<String> xDatas) {
        this.xDatas = xDatas;

        invalidate();
    }

}

File: base/src/main/java/io/moka/lib/views/SpreadTextView.kt
package io.moka.lib.views


import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View

import java.util.ArrayList


class SpreadTextView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {

    private var textPaint: TextPaint? = null
    private var textBoundRect: Rect? = null

    private var firstX: Float = 0.toFloat()

    private var dayY: Float = 0.toFloat()

    private var xDatas = ArrayList<SpreadData>()

    init {
        initView()
    }

    private fun initView() {
        textBoundRect = Rect()

        textPaint = TextPaint()
        textPaint!!.textAlign = Paint.Align.CENTER
        textPaint!!.isAntiAlias = true

        xDatas.add(SpreadData("월"))
        xDatas.add(SpreadData("화"))
        xDatas.add(SpreadData("수"))
        xDatas.add(SpreadData("목"))
        xDatas.add(SpreadData("금"))
        xDatas.add(SpreadData("토"))
        xDatas.add(SpreadData("일"))
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        val dayWidth = w / 7.0f
        dayY = h / 2.0f

        firstX = dayWidth / 2

        val textSize = h / 2.5f
        textPaint!!.textSize = textSize
    }

    override fun onDraw(canvas: Canvas) {
        xDatas.forEachIndexed { index, spreadData ->
            textPaint!!.color = spreadData.color
            if (spreadData.isBold)
                textPaint!!.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
            else
                textPaint!!.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL)

            drawText(canvas, spreadData.text, textPaint, firstX + (firstX * 2 * index), dayY)
        }
    }

    private fun drawText(canvas: Canvas, text: String, textPaint: TextPaint?, centerX: Float, centerY: Float) {
        textPaint!!.getTextBounds(text, 0, text.length, textBoundRect)
        canvas.drawText(text, centerX, (centerY + textBoundRect!!.height() / 2).toInt().toFloat(), textPaint)
    }

    /*
    Setter
     */

    fun setXDatas(xDatas: ArrayList<SpreadData>) {
        this.xDatas = xDatas

        invalidate()
    }

}
